p2p/transport/webrtc/udpmux/packetbuffer.go

```diff
@@ -0,0 +1,95 @@
+package udpmux
+
+import (
+	"context"
+	"io"
+	"net"
+	"sync"
+
+	pool "github.com/libp2p/go-buffer-pool"
+)
+
+type packet struct {
+	addr net.Addr
+	size int
+}
+
+type packetBuffer struct {
+	ctx         context.Context
+	mu          sync.Mutex
+	pdata       *pool.Buffer
+	pkts        []packet
+	notify      chan struct{}
+	readWaiting bool
+}
+
+func newPacketBuffer(ctx context.Context) *packetBuffer {
+	return &packetBuffer{
+		ctx:    ctx,
+		pdata:  new(pool.Buffer),
+		notify: make(chan struct{}),
+	}
+}
+
+func (pb *packetBuffer) readFrom(buf []byte) (int, net.Addr, error) {
+	for {
+		pb.mu.Lock()
+
+		if len(pb.pkts) != 0 {
+			pkt := pb.pkts[0]
+			pktdata := pb.pdata.Next(pkt.size)
+			pb.pkts = pb.pkts[1:]
+			copied := copy(buf, pktdata)
+			if copied < len(pktdata) {
+				pb.mu.Unlock()
+				return copied, pkt.addr, io.ErrShortBuffer
+			}
+
+			pb.mu.Unlock()
+			return copied, pkt.addr, nil
+		}
+
+		notify := pb.notify
+
+		pb.readWaiting = true
+		pb.mu.Unlock()
+
+		select {
+		case <-notify:
+		case <-pb.ctx.Done():
+			pb.mu.Lock()
+			if len(pb.pkts) == 0 {
+				pb.pdata.Reset()
+				pb.mu.Unlock()
+				return 0, nil, io.EOF
+			}
+			pb.mu.Unlock()
+		}
+	}
+}
+
+func (pb *packetBuffer) writePacket(buf []byte, addr net.Addr) error {
```

Comment:
packetBuffer doesn’t sound like an appropriate name for something that writes packets. Is this the a connection?

Reply:
No, this only pushes to the packet buffer. I have renamed these.